ned/instance.py

```diff
@@ -150,42 +151,66 @@ def _ec2_config(
     return definition
 
 
-class Barrier:
+class InstanceBarrier:
     def wait(self, inst: Instance) -> Instance:
         return inst
 
 
 @dataclass(frozen=True)
-class InstanceRunningBarrier(Barrier):
+class InstanceReadyBarrier(InstanceBarrier):
     ec2: botocore.client.BaseClient
-    sleep_time: float = 10
-    retry_count: int = 9
+    sleep_time: float = 5
+    retry_count: int = 20
+
+    def _is_ready(self, inst: Instance) -> bool:
+        if inst.state != "running":
+            self._warn(f'Instance state is "{inst.state}", waiting for "running"')
+            return False
+
+        try:
+            resolved_ip = socket.gethostbyname(inst.public_dns_name)
+        except socket.gaierror as exc:
+            self._warn(f"Host lookup error: {exc}")
+            return False
+
+        if inst.public_ip_address != resolved_ip:
```

Comment:
Check if the DNS resolution `inst.public_dns_name` gives us the expected IP. If not, the instance is not ready.

I wasn't able to get it to live fail on this, but it's still worth checking to be sure.